### PR TITLE
[JENKINS-62024] Add quick filter box for dropdown lists in hetero-list

### PR DESCRIPTION
**Pull request by johnlinp**
See [JENKINS-62024](https://issues.jenkins-ci.org/browse/JENKINS-62024).

image

image

### Proposed changelog entries

* Add quick filter box for menus created by buttons.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention Oleg Nenashev

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by daniel-beck** [comment]
Reviewer note: This change will not be reflected in the PR tester by default.

**Comment by fqueiruga** [comment]
I don't think modifying the YUI code is a good approach to this. It's library code that's been stable for 12+ years, so I'd be very wary when making the decision to touch it.

The DragDrop widget also builds on the YUI code, but does so by decorating it: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/resources/lib/form/dragdrop/dragdrop.js. This, if possible, is the approach I suggest. Otherwise I'd just build a new dropdown widget altogether.

**Comment by daniel-beck** [comment]
> I don't think modifying the YUI code is a good approach.

We've done it before 🤷  It's not like the library is pristine.

**Comment by fqueiruga** [comment]
Still, I'd suggest trying the other approach first.

**Comment by johnlinp** [comment]
@fqueiruga I was looking for an example of decorations on YUI but failed. Thank you for providing the DragDrop example. I'll work on it.

**Comment by johnlinp** [comment]
Hi @fqueiruga,

I pushed a new commit that wraps around `YAHOO.widget.Button` and apply the quick filter box on it. Please take a look at it if it's acceptable. Thank you.

**Comment by varyvol** [comment]
@johnlinp I don't feel confident enough (with my UI knowledge) to approve, but it looks good. Maybe you could add some explanatory comments to the code and some screenshots where the functionality can be observed.

@fqueiruga could you maybe have a look? The current status uses your preferred approach.

**Comment by johnlinp** [comment]
Hi @fqueiruga,

Would you please review my PR? Thanks!

**Comment by fqueiruga** [comment]
I think the approach is good even though the styling is rough.

Did you consoider only applying the filter when the number of options is over 5 or something? There are many hetero lists with only 2 options and that may be overkill.

**Comment by johnlinp** [comment]
Hi @fqueiruga,

Thank you for the review. Yes, I think it makes sense to only apply the filter when the number of options is over 5. I'll push a new commit to implement this.

**Comment by johnlinp** [comment]
@alecharp Yes, I would love to continue working on it. Let's see if the reviewers have any comments. Thanks.

**Comment by janfaracik** [comment]
Works well 👍 UI is a touch rough but so is YUI in general. 

I want to update/replace the YUI menus eventually but that's no reason to delay this enhancement.

Only thing I'd say is see if there's a way of disabling autocomplete/autocorrect on the filter field.

**Comment by alecharp** [comment]
@johnlinp this is great. Could you work on the merged conflicts so we can move this PR forward. Thanks.

**Comment by johnlinp** [comment]
@alecharp I've updated the branch to resolve the conflict. Thanks.

@janfaracik I didn't see autocomplete or autocorrect on my end. Can you provide the reproduction steps to show autocomplete or autocorrect for the filter? Thank you!

**Comment by alecharp** [comment]
@janfaracik I saw in a PR that you are removing the YUI tooltip implementation. Is there an initiative to remove YUI completely from the core? The work here from @johnlinp seems to be good and could be integrated, no? WDYT @fqueiruga ?

File: core/src/main/resources/lib/form/filter-menu-button/filter-menu-button.js
Comment (janfaracik) [inline core/src/main/resources/lib/form/filter-menu-button/filter-menu-button.js:30]:
```suggestion
    filterInput.setAttribute("placeholder", "Filter");
    filterInput.setAttribute("spellcheck", "false");
    filterInput.setAttribute("type", "search");
```

This fixed the spellcheck problem for me, also uses the correct input type so you get a nice little ⓧ button for clearing text 👍

**Comment by janfaracik** [comment]
> @janfaracik I saw in a PR that you are removing the YUI tooltip implementation. Is there an initiative to remove YUI completely from the core? The work here from @johnlinp seems to be good and could be integrated, no? WDYT @fqueiruga ?

Hoping to remove as much of the YUI stuff as possible eventually, might take a good while though just due to how much there is. This work is definitely good though and should be integrated 👍

File: core/src/main/resources/lib/form/filter-menu-button/filter-menu-button.js
Comment (fqueiruga) [inline core/src/main/resources/lib/form/filter-menu-button/filter-menu-button.js:1]: If these functions are to be global I'd suggest maybe namespacing them. I can see `createFilterMenuButton` being a name that could be easily overriden.

**Comment by fqueiruga** [comment]
My comments about holding off removing YUI are 2 years old and were of a time before @janfaracik 's amazing work. I'd say that, if possible, it may be beneficial to include this in YUI removal efforts.

**Comment by basil** [comment]
I am getting a generally positive assessment of this PR from reviewers, but on the other hand none of them have approved the PR, which makes me wonder why. Reviewers should consider providing an [Apache-style vote](https://www.apache.org/foundation/voting.html) (+1, 0, -1, and fractions) to make their assessment explicit. If the general consensus is strongly positive (e.g., explicit GitHub approvals or +1 votes), I will merge the PR. In contrast, if the general consensus is strongly negative (which I do not believe is the case for this PR!) I would close the PR. If, on the other hand, the general consensus is closer toward neutral, it would be helpful for reviewers to explicitly state which action items are needed in order to move the PR closer to the positive side of the voting number line.

**Comment by timja** [comment]
I like the change, 2 areas I think we can improve before I'll give a +1

1. Focus state to match other components, e.g.:

image

2. Text should be aligned to other text not touching the left edge

image

File: core/src/main/resources/lib/form/filter-menu-button/filter-menu-button.js
Comment (timja) [inline core/src/main/resources/lib/form/filter-menu-button/filter-menu-button.js:34]: changed to `input` from `keyup` so that the search clear event works

**Review by timja — APPROVED** [review]

File: core/src/main/resources/lib/form/filter-menu-button/filter-menu-button.js
Comment (timja) [reply core/src/main/resources/lib/form/filter-menu-button/filter-menu-button.js]: it's already quite a long name, I think it will be okay for now anyway. Unless you've got any specific suggestions?

**Comment by timja** [comment]
Pushed a commit to address my concerns. (also reformatted it to match editorconfig file)

**Review by NotMyFault — APPROVED** [review]
Looks ok to me, this component (likely) needs to be revisited in Jan's anyway, to align it with the overhauled look of the project configuration view.

Though to note, various tests are failing.

**Comment by timja** [comment]
I assume that was an issue with optional chaining and html unit =/

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by johnlinp** [comment]
Thank you @timja for reviewing my code and updating my branch!

**Pull request merged**
